### ISSUE TITLE
A space-separated `disallowed-tools:` is read as one token, so a working fence audits as closing nothing

### ISSUE THREAD

**Issue by vlad-ryzhkov** [zernie/agent-lint#217]
## Summary

`disallowed-tools:` written as a **space-separated string** is read as one token, so a fence that Claude Code honours at runtime audits as `fence: "ineffective"` — "closes no lethal-trifecta leg". The comma-separated string and the YAML list are both read correctly.

Claude Code documents all three forms as equivalent. From [the skills reference](https://code.claude.com/docs/en/skills):

> Accepts a space- or comma-separated string, or a YAML list.

Its own example for `allowed-tools` is space-separated *and* contains spaces inside the tokens:

```yaml
allowed-tools: Bash(git add *) Bash(git commit *) Bash(git status *)
```

So an author can write a correct, working fence and see the audit tell them it closes nothing.

## Reproduction

`vigiles@26.1.1`, harness of 22 skills, `claude-code` dialect. One skill (`spec-audit`) changed, one variable — the separator. Same three tool names in all three runs.

| `disallowed-tools:` value | Safety | `spec-audit` flagged as ineffective? |
| --- | --- | --- |
| `WebFetch, WebSearch, Bash` | **81** | no |
| `WebFetch WebSearch Bash` | **80** | **yes** |
| YAML list (`- WebFetch` / `- WebSearch` / `- Bash`) | **81** | no |

Reverting the separator restored 81, so the effect is reversible and attributable to that one character class.

Minimal case:

```yaml
---
name: spec-audit
allowed-tools: "Read Write Glob"
disallowed-tools: WebFetch WebSearch Bash
---
```

`npx vigiles audit` reports:

> `disallowed-tools: WebFetch WebSearch Bash` closes no lethal-trifecta leg — a leg is closed only when EVERY built-in that supplies it is denied. Private-data read is still supplied by Read, Grep, Glob, Bash, untrusted intake by WebFetch, WebSearch, Bash, exfiltration by WebFetch, WebSearch, Bash.

Every supplier of the exfiltration leg *is* named. The list just never got split.

## Root cause

`dist/core/frontmatter-read.js`, `splitList()` at line 135 splits on `,` only:

```js
function splitList(raw) {
    return raw
        .replace(/^\[|\]$/g, "")
        .split(",")
        .map((t) => t.trim().replace(/^["']|["']$/g, ""))
        .filter((t) => t.length > 0);
}
```

`"WebFetch WebSearch Bash"` therefore yields `["WebFetch WebSearch Bash"]` — a single token matching no built-in name. The `Array.isArray(v)` branch in `frontmatterList()` is why the YAML-list form works, and the comma path is why the documented comma form works.

The doc comment above `frontmatterList()` already describes the field as "an array or a comma list", so this reads as a known-narrow reader rather than an oversight — but Claude Code accepts a third form, and that form silently produces the opposite verdict.

## Why the one-line fix is wrong

`raw.split(/[,\s]+/)` would break tool patterns, which legitimately contain spaces:

```yaml
disallowed-tools: Bash(git push *), WebFetch
```

Splitting that on whitespace gives `Bash(git`, `push`, `*)`, `WebFetch` — three garbage tokens plus one good one. Worse, it would silently turn a bounded grant into an unrecognised one, and `bashGrantIsUnbounded()` returns `true` when in doubt, so the harness would flip from clean to exposed for a reason the author cannot see. That is the same "false safe vs false exposed" trade already reasoned about in `core/lethal-trifecta.js`, landing on the bad side.

A tokenizer needs to treat whitespace as a separator only at paren depth 0:

```js
function splitList(raw) {
    const s = raw.trim().replace(/^\[|\]$/g, "");
    const out = [];
    let buf = "", depth = 0;
    for (const ch of s) {
        if (ch === "(") depth++;
        else if (ch === ")") depth = Math.max(0, depth - 1);
        if (depth === 0 && (ch === "," || /\s/.test(ch))) {
            if (buf.trim()) out.push(buf.trim());
            buf = "";
            continue;
        }
        buf += ch;
    }
    if (buf.trim()) out.push(buf.trim());
    return out.map((t) => t.replace(/^["']|["']$/g, "")).filter((t) => t.length > 0);
}
```

Cases worth pinning in a test, all from the Claude Code docs or its grammar:

| input | expected |
| --- | --- |
| `WebFetch, WebSearch` | `["WebFetch","WebSearch"]` |
| `WebFetch WebSearch` | `["WebFetch","WebSearch"]` |
| `Bash(git add *) Bash(git commit *)` | `["Bash(git add *)","Bash(git commit *)"]` |
| `Bash(git push *), WebFetch` | `["Bash(git push *)","WebFetch"]` |
| `"Read Write Glob"` | `["Read","Write","Glob"]` |
| `[Read, Grep]` | `["Read","Grep"]` |

All seven rows above (plus `WebFetch,  WebSearch ,Bash` for irregular spacing) pass against that function — I ran them before posting rather than eyeballing them.

Note the quoted-string case: `allowed-tools: "Read Write Glob"` is common in the wild — it is what every skill in my harness had before this change. Whether YAML strips the quotes before the reader sees the value or the salvage path hands them through, the result is a single space-separated string on the same code path.

## Impact

Silent and in the unsafe direction for the author's understanding: they wrote a fence, Claude Code enforces it, and `audit` tells them it closes nothing. The likely reactions are to add tools that are already denied, or to conclude the metric is noise. `allowed-tools` goes through the same `splitList`, so a space-separated allow list is also read as one token — I have not worked out whether anything downstream depends on that, but it is the same code path.

## Environment

- `vigiles@26.1.1` via `npx`
- macOS 26.6.2 arm64, Node from Homebrew
- Harness: 22 skills, hand-written `CLAUDE.md`, no spec, `claude-code` dialect
- Repo the numbers came from: <https://github.com/vlad-ryzhkov/ai-context-engineering-for-qa>

Happy to send a PR with the tokenizer plus the table above as test cases if you want it in that shape.

**Comment by zernie** [comment]
Fixed and merged in #221 (`c4e75ae`). Thank you — this was an unusually good report, and it was right on every point.

### Reproduced before changing anything

Three of the seven shapes came back as one token:

| input | before |
| --- | --- |
| `WebFetch WebSearch` | `["WebFetch WebSearch"]` |
| `Bash(git add *) Bash(git commit *)` | one token |
| `"Read Write Glob"` | one token |

The quoted row is why this stayed invisible for so long — a YAML-quoted scalar whose *value* is a space-separated list is what most skills in the wild actually write, exactly as you said.

### Your tokenizer shipped as proposed

Your six rows plus irregular spacing and an unbalanced paren are the tests, and you're credited as `Co-Authored-By` on the commit. Your reasoning against `split(/[,\s]+/)` was the load-bearing part of the report: it would have traded a false-clean verdict for a false-**exposed** one through `bashGrantIsUnbounded()`, and that is the worse direction to fail in.

### Your open question has an answer, and it is worse than the bug you reported

You wrote: *"`allowed-tools` goes through the same `splitList` — I have not worked out whether anything downstream depends on that."* It does.

A **skill's** `allowed-tools:` only pre-approves, so there a mis-split misleads the audit. But a **subagent's** `tools:` is the allowlist the PreToolUse rail actually *denies against*, and it reads the same function. Measured on `tools: Read Grep Glob`, before the fix:

```
Read  DENY   ← explicitly granted
Grep  DENY   ← explicitly granted
Bash  DENY
```

The joined token matched nothing, so the agent was denied **every** tool it had been given. That's a broken runtime rail rather than a misleading report, and it now has its own mutation-proven test.

### One case neither of us had

A review bot caught a gap in the first cut: a literal *unmatched* paren inside a grant's own command (`Bash(printf '( %s' foo) Read`) left the depth counter at 1 and swallowed the rest. Quote state now gates the depth counter only — never the splitting, since suppressing splitting inside quotes would reintroduce your original bug for every skill that quotes its list.

You offered to send the PR and I went ahead instead — sorry for jumping it. Please do send the next one; reports at this level of care are rare, and the two extra surfaces only turned up because yours was precise enough to follow.

---
_Generated by [Claude Code](https://claude.ai/code)_
